<a href="https://colab.research.google.com/github/mukteshsingh95/Play-Store-App/blob/main/Play_Store_App.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# importing libraries
# numpy-for linear transformation
# pandas-for data manipulation and analysis
# matplotlib,seaborn,plotly-for visualization
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
#Mounting the drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [8]:
#loading csv files
playstore_data_path = '/content/drive/MyDrive/DATA/Play Store Data (1).csv'
user_review_data_path = '/content/drive/MyDrive/DATA/User Reviews (1).csv'
playstore_df = pd.read_csv(playstore_data_path)
review_df = pd.read_csv(user_review_data_path)

# **Exploring Playstore and User Review Dataframe.**

**1**.**Playstore** **Dataframe**

In [10]:
#shape of the play store dataset
playstore_df.shape

(10841, 13)

In [12]:
#displaying first 5 rows of the dataset
playstore_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [11]:
#displaying last 5 rows of the dataset
playstore_df.tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [13]:
# information about the features in the dataset
playstore_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [14]:
#shape of the Review dataset
review_df.shape

(64295, 5)

In [15]:
#displaying first 5 rows of the dataset
review_df.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

In [16]:
#displaying last 5 rows of the dataset
review_df.tail()

App Translated_Review Sentiment  \
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN

In [17]:
# information about the features in the dataset
review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


# **Lets explore all the features in the Dataset**

**App** - describes the name of each apps.

**Category** - Category at which the app belongs.


**Rating** -Average rating of the app received from its users.

**Reviews** - The total number of reviews got by the apps from its users.

**Size**- Memory size occupied by the app in the mobile device.

**Installs**- The total number of downloads for the application.

**Type** - whether the app is free or paid

**Price**-If the app is paid, what is the cost required to install the app.

**Content Rating** - It specifies weather the app is suitable for all age group or not.

**Genres**- the other categories to which the app can belong.

**Last Updated**- Last updated date of the app.

**Current Ver**-The current version of the app.

**Android Ver**-The Android Version which can support the application.

# **2.Review Dataframe**

In [18]:
#shape of the Review dataset
review_df.shape

(64295, 5)

In [19]:
#displaying first 5 rows of the dataset
review_df.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

In [20]:
#displaying last 5 rows of the dataset
review_df.tail()

App Translated_Review Sentiment  \
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN

In [21]:
# information about the features in the dataset
review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


# **Lets explore all the features in the Dataset**

1. **App** - describes the name of each apps.

2. **Translated_Review**-English translation of the user reviews.

3. **Sentiment** - It gives the emotion of the reviewer related with his review.It can be ‘Positive’, ‘Negative’, or ‘Neutral’.
Sentiment_Polarity - The polarity of the review.It ranges from [-1 to 1]. -1 means "negative sentiment" and 1 means "positive sentiment"

4. **Sentiment_Subjectivity** - How the opinion of a particular reviewer is aligned with the opinion of the general public.It ranges from [0 to 1]. Heigher the subjectivity means the review is closer to the opinion of general public.

# **Data Cleaning**

# **Play Store Dataset**

In [22]:
#Making copy of playstore dataframe
df = playstore_df.copy()

# **1- Converting "Review" column type from "object" to "int".**

In [24]:
df['Reviews']=df['Reviews'].apply(lambda x : eval(x))

SyntaxError: ignored

 **While converting Reviews into "int" type we found some Error. Which show that one of the value in review column is 3.0M.So we have to remove that row from the data set.**

In [25]:
# Finding the row index of that value.
df[df['Reviews']=='3.0M']

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

 **We have found that index no 10472 contain the value 3.0M.**

In [26]:
# Dropping that index value using drop method and reseting index again.
df = df.drop(10472).reset_index(drop=True)

In [27]:
# Evaluating the values.
df['Reviews']=df['Reviews'].apply(lambda x : eval(x))

# **2- Converting "Size" column type from "object" to "float".**

In [28]:
#printing last 5 rows of dataset
df.tail()

App             Category  \
10835                               Sya9a Maroc - FR               FAMILY   
10836               Fr. Mike Schmitz Audio Teachings               FAMILY   
10837                         Parkinson Exercices FR              MEDICAL   
10838                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10839  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
10835     4.5       38                 53M       5,000+  Free     0   
10836     5.0        4                3.6M         100+  Free     0   
10837     NaN        3                9.5M       1,000+  Free     0   
10838     4.5      114  Varies with device       1,000+  Free     0   
10839     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10835       Everyone          Education     July 25, 2017                1.48   
10836       Everyone          Education      July 6, 2018                 1.0   
10837       Everyone            Medical  January 20, 2017                 1.0   
10838     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10839       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10835          4.1 and up  
10836          4.1 and up  
10837          2.2 and up  
10838  Varies with device  
10839  Varies with device

# It is observed that "Size column contain values in MB (M), kb(k) and "Varies with device".





In [29]:
# Checking the shape of size column which is equal to Varies with device
df[df['Size']=='Varies with device'].shape

(1695, 13)

In the size coloumn the sizes are in different units. Some of them are in 'M' means MB some of them are in 'K' means KB.For those which are in MB we just convert it to float after removing the 'M' charecter from it. For those which are in KB first we convert it to float after removing 'K from it than we convert it to MB by dividing it by **1024**

Again we can see there are **1695 rows** that has "**Varies with device"** value in **"Size"** column. Replacing this value with mean value will affect the visualization. Replacing Varies with device value with np.NaN which is of float type.

In [30]:
def converting_size_into_float(string):
  '''
  This function helps in removing 'M'(MB) and 'k'(kb) which are present at 
  the end of the string and replace 'Varies with device' with np.NaN.
   This function also evaluate the values present. 
  '''
  
  if string[-1] == 'M':
    return eval(string.strip('M'))

  elif string[-1] == 'k':
    a =string.strip('k')
    b = str(round(eval(a)/1024,1))
    return eval(b)

  elif string == 'Varies with device':
    string = np.NaN  
    return string
  else:
    return eval(string)

In [31]:
# Applying defined function.
df['Size_in_MB']= df['Size'].apply(lambda x : converting_size_into_float(x))

# 3- Converting "Install" column type from "object to "int".

In **Install** column values are of **object** type and contain '+' and ',' in them. So we are going remove '+' and ',' from the values and then convert them into **int** type using **eval** method.



In [32]:
# Creating function to remove + and ,.
def remove_plus_and_comma(string):
  '''
  This function removes '+' and ',' from the string.
  '''
  string = string.replace(',','')
  string = string.strip('+')
  return string

In [33]:
# Applying defined function on the column and evaluating those values.
df['Installs'] = df['Installs'].apply(lambda x: eval(remove_plus_and_comma(x)))

# 4- Converting **"Price"** column type from **"object"** to **"float"**.
**Price** column value has $ symbol in them and they are of object type. we'll remove the symbol and change the type.

In [34]:
# Creating function to remove $ symbol.
def remove_sign(string):
  '''
  This function removes $ symbol from the string and convert given string data type from 'str' to 'float'.
  '''
  return round(float(string.strip('$')),2)

In [35]:
# Applying the function.
df['price_in_dollar'] = df['Price'].apply(lambda x : remove_sign(x))

### 5- Converting **Last Update** type from **"object"** to **"datetime"**.

In [36]:
# Converting str into datetime formate using 'to_datetime' function.
df['Last Updated'] = pd.to_datetime(df['Last Updated'])

# **Handling the NaN values**

In [37]:
# Total number of null values in the dataframe 
df.isnull().sum()

App                   0
Category              0
Rating             1474
Reviews               0
Size                  0
Installs              0
Type                  1
Price                 0
Content Rating        0
Genres                0
Last Updated          0
Current Ver           8
Android Ver           2
Size_in_MB         1695
price_in_dollar       0
dtype: int64

# 6- Handling Rating column

In [38]:
#checking the null value in Rating colunn
len(df[df['Rating'].isnull()]['Rating'])

1474

# Rating column contain 1474 NaN values. We cannot drop this much amount of row from dataset because it will lose a huge percentage of information.We gonna replace all the NaN values with the average of non-null values.

In [39]:
# Finding average of non-null values from Rating column.
non_null_mean= round(df[~df['Rating'].isnull()]['Rating'].mean(),1)

In [40]:
# Replacing null values with average rating.
df['Rating'].fillna(value= non_null_mean, inplace=True)

# 7- Removing Null value from "Type" column.

**Type** column contain only one NaN value.

In [41]:
# checking the null value in Type colunn
df[df['Type'].isnull()]['Type']

9148    NaN
Name: Type, dtype: object

It is observed that Type column contain Null value at index 9148.

In [42]:
# Removing row having index 9148 and reseting the index.
df = df.drop(9148).reset_index(drop=True)